File: leetcode/leetcode_cpp/maximum-number-of-vowels-in-a-substring-of-given-length.cpp
/*
https://leetcode.com/problems/maximum-number-of-vowels-in-a-substring-of-given-length

1. Sliding Window
time: o(n)
space: o(1)
*/
class Solution {
public:
    int maxVowels(string s, int k) {
        unordered_set<char> c {'a', 'e', 'i', 'o', 'u'};
        int n = s.size(), r = 0;
        int cnt = 0, res = 0;
        while (r < n) {
            if (c.count(s[r])) cnt += 1;
            if (r >= k && c.count(s[r-k])) cnt -= 1;
            r += 1;

            res = max(res, cnt);
        }
        return res;
    }
};

File: leetcode/leetcode_py/maximum-number-of-vowels-in-a-substring-of-given-length.py
'''
https://leetcode.com/problems/maximum-number-of-vowels-in-a-substring-of-given-length

1. Sliding Window
time: o(n)
space: o(1)
'''
class Solution:
    def maxVowels(self, s: str, k: int) -> int:
        vowels = {'a', 'e', 'i', 'o', 'u'}
        n = len(s)
        res, cnt = 0, 0
        for i in range(n):
            if i >= k and s[i-k] in vowels:
                cnt -= 1
            if s[i] in vowels:
                cnt += 1
            res = max(res, cnt)
        return res

        